File: figures/07-hash-tables/snippet-open-address-hash-table.py
class OpenAddressHashTable:
    _MAX_PROBING_LOAD = 0.65

    def __init__(self, capacity=8):
        self._keys = [None] * capacity
        self._values = [None] * capacity
        self._states = [EMPTY] * capacity
        self._size = 0
        self._tombstones = 0

    def __len__(self):
        return self._size

    def capacity(self):
        return len(self._keys)

    def tombstones(self):
        return self._tombstones

    def _home(self, key):
        return key % len(self._keys)

    def _find_index(self, key):
        m = len(self._keys)
        start = self._home(key)
        for i in range(m):
            j = (start + i) % m
            if self._states[j] == EMPTY:
                return -1
            if self._states[j] == OCCUPIED and self._keys[j] == key:
                return j
        return -1

    def get(self, key):
        j = self._find_index(key)
        return None if j < 0 else self._values[j]

    def put(self, key, value):
        if (self._size + self._tombstones + 1) / len(self._keys) > self._MAX_PROBING_LOAD:
            self._resize(len(self._keys) * 2)
        m = len(self._keys)
        start = self._home(key)
        first_deleted = -1
        for i in range(m):
            j = (start + i) % m
            if self._states[j] == OCCUPIED and self._keys[j] == key:
                old = self._values[j]
                self._values[j] = value
                return old
            if self._states[j] == DELETED and first_deleted < 0:
                first_deleted = j
            if self._states[j] == EMPTY:
                target = first_deleted if first_deleted >= 0 else j
                self._place(target, key, value)
                return None
        self._place(first_deleted, key, value)
        return None

    def _place(self, j, key, value):
        if self._states[j] == DELETED:
            self._tombstones -= 1
        self._keys[j] = key
        self._values[j] = value
        self._states[j] = OCCUPIED
        self._size += 1

    def remove(self, key):
        j = self._find_index(key)
        if j < 0:
            return None
        old = self._values[j]
        self._keys[j] = self._values[j] = None
        self._states[j] = DELETED
        self._size -= 1
        self._tombstones += 1
        return old

    def _resize(self, new_capacity):
        old_keys, old_values, old_states = self._keys, self._values, self._states
        self._keys = [None] * new_capacity
        self._values = [None] * new_capacity
        self._states = [EMPTY] * new_capacity
        self._size = self._tombstones = 0
        for k, v, s in zip(old_keys, old_values, old_states):
            if s == OCCUPIED:
                self.put(k, v)

File: figures/07-hash-tables/snippet-open-address-hash-table.java
    private int[] keys, values;
    private byte[] states;
    private int size, tombstones;

    public OpenAddressHashMap(int capacity) {
        keys = new int[capacity];
        values = new int[capacity];
        states = new byte[capacity];
    }

    private int home(int key) { return Math.floorMod(key, keys.length); }

    public int size() { return size; }

    private int findIndex(int key) {
        int start = home(key);
        for (int i = 0; i < keys.length; i++) {
            int j = (start + i) % keys.length;
            if (states[j] == EMPTY) return -1;
            if (states[j] == OCCUPIED && keys[j] == key) return j;
        }
        return -1;
    }

    public Integer get(int key) {
        int j = findIndex(key);
        return j < 0 ? null : values[j];
    }

    public Integer put(int key, int value) {
        if ((size + tombstones + 1.0) / keys.length > MAX_PROBING_LOAD) resize(keys.length * 2);

        int start = home(key), firstDeleted = -1;
        for (int i = 0; i < keys.length; i++) {
            int j = (start + i) % keys.length;
            if (states[j] == OCCUPIED && keys[j] == key) {
                int previous = values[j];
                values[j] = value;
                return previous;
            }
            if (states[j] == DELETED && firstDeleted < 0) firstDeleted = j;
            if (states[j] == EMPTY) {
                place(firstDeleted >= 0 ? firstDeleted : j, key, value);
                return null;
            }
        }
        place(firstDeleted, key, value);
        return null;
    }

    private void place(int j, int key, int value) {
        if (states[j] == DELETED) tombstones--;
        keys[j] = key;
        values[j] = value;
        states[j] = OCCUPIED;
        size++;
    }

    public Integer remove(int key) {
        int j = findIndex(key);
        if (j < 0) return null;
        int previous = values[j];
        states[j] = DELETED;
        size--;
        tombstones++;
        return previous;
    }

    private void resize(int newCapacity) {
        int[] oldKeys = keys, oldValues = values;
        byte[] oldStates = states;
        keys = new int[newCapacity];
        values = new int[newCapacity];
        states = new byte[newCapacity];
        size = tombstones = 0;
        for (int i = 0; i < oldKeys.length; i++)
            if (oldStates[i] == OCCUPIED) put(oldKeys[i], oldValues[i]);
    }
